source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -190,6 +190,8 @@ Filter::~Filter() {
   // Disable access log flush timer if it is enabled.
   disableAccessLogFlushTimer();
 
+  getStreamInfo().setStreamState(StreamInfo::StreamState::Ended);
```

Comment:
I think the responsibility to start/end L4 streams should be owned by the listeners, not by the tcp_proxy. For example, HCM L4 streams doesn't will always have the stream state unset, which hurts L4 stream logging for HTTP connections. Can we refactor this so that the listener filter manager sets start/end at the right places? In-progress seems to be application-specific - so that's OK to leave in the proxy itself, although a simpler transition for HCM L4 stream to in-progress would be better.

Reply:
Thanks for the quick response!

HCM is a network proxy like tcp_proxy so it has its own TCP connection stream infos to manage. There's a listener-level access log that could be used in combination with HCM. 

I understand the need for InProgress state - that is indeed application specific (e.g. upstream connection for tcp_proxy or 200 for HCM). But "ended" seems like a universal state so it could be done uniformly at a layer above tcp_proxy, hence, the listener recommendation.

"Started" is a bit unclear - I'd expect it means that something has happened, e.g. bytes have been consumed or produced - and that happens earlier than tcp_proxy, somewhere in the transport socket.

Reply:
> "Started" is a bit unclear - I'd expect it means that something has happened, e.g. bytes have been consumed or produced - and that happens earlier than tcp_proxy, somewhere in the transport socket.

Yeah so it means that an upstream connection was made for tcp_proxy, and request headers have been recieved for HCM. I think the added docs explains this, but if it's not clear enough please lmk and I'll try to make it clearer. 
Anyways, I'm not sure that I'm convinced about separating some of the implementation to listener level and some to proxy level, to me it makes more sense as it is right now (minus maybe elaborating more in docs if needdd) but I don't want to be too hard opinionated here. Maybe @paul-r-gall can add his opinion on this?

Reply:
Ok, given that Started/InProgress depends on the details of the proxy, we can only guarantee Ended at the listener level, so it's not useful to do it now.